ssl/ssl_lib.cc

```diff
@@ -3311,33 +3310,16 @@ int SSL_set1_curves_list(SSL *ssl, const char *curves) {
 }
 
 size_t SSL_client_hello_get0_ciphers(SSL *ssl, const unsigned char **out) {
-  STACK_OF(SSL_CIPHER) *client_cipher_suites = SSL_get_client_ciphers(ssl);
-  if (client_cipher_suites == nullptr) {
+  if (SSL_get_client_ciphers(ssl) == nullptr) {
       return 0;
   }
+  const char * ciphers = ssl->all_client_cipher_suites.get();
+  if (ciphers == nullptr) {
+    return 0;
+  }
 
-  size_t num_ciphers = sk_SSL_CIPHER_num(client_cipher_suites);
   if (out != nullptr) {
-    // Hasn't been called before
-    if(ssl->client_cipher_suites_arr.empty()) {
-      if (!ssl->client_cipher_suites_arr.Init(num_ciphers)) {
-        OPENSSL_PUT_ERROR(SSL, ERR_R_MALLOC_FAILURE);
-        return 0;
-      }
-
-      // Construct list of cipherIDs
-      for (size_t i = 0; i < num_ciphers; i++) {
-        const SSL_CIPHER *cipher = sk_SSL_CIPHER_value(client_cipher_suites, i);
-        uint16_t iana_id = SSL_CIPHER_get_protocol_id(cipher);
-
-        CRYPTO_store_u16_be(&ssl->client_cipher_suites_arr[i], iana_id);
-      }
-    }
-
-    assert(ssl->client_cipher_suites_arr.size() == num_ciphers);
-    *out = reinterpret_cast<unsigned char*>(ssl->client_cipher_suites_arr.data());
+    *out = reinterpret_cast<const unsigned char*>(ciphers);
   }
-
-  // Return the size
-  return num_ciphers;
+  return strlen(ciphers);
```

Comment:
ahh, the size+1 allocation was needed b/c the value is being interpreted as a null-terminated string here. I think it'd be better to explicitly store the size of this array and use that here.

Reply:
Do we know that the cipher content cannot contain a 0-byte?

Reply:
Hmmm that's an interesting case. If the cipher data contains valid cipher IDs, then we know. But I guess a client could hypothetically send invalid cipher suites that contain the null byte - investigating.

Reply:
This is a problem from the OpenSSL API, because lots of cipher suites include a 0 byte for the first part of the ID, and the all null cipher suite is 0x0, 0x0 https://datatracker.ietf.org/doc/html/rfc5246#appendix-A.5. However reading https://www.openssl.org/docs/man1.1.1/man3/SSL_client_hello_get0_ciphers.html more closely it also returns the length of the array, so I don't think we need to null terminate this string because OpenSSL also isn't treating it as a string. But we do need to track the length.

Reply:
So we can directly return the underlying buffer and don't have to worry about any scrap data at the end that was never zeroed out. It becomes the caller's responsibility to read the right number of bytes using the return value of get0_ciphers?

Reply:
Yes, that's how I'm reading the OpenSSL docs.

Reply:
Got it, that will be the new behavior after this PR!